File: react-native/src/Scheduler.ts
/**
 * Edge Veda SDK - Scheduler
 * Priority-based task scheduler with budget enforcement.
 *
 * The Scheduler manages concurrent inference workloads, enforces EdgeVedaBudget
 * constraints, and emits BudgetViolation events when limits are exceeded.
 *
 * JavaScript is single-threaded in React Native, so no mutex is needed.
 * Async/await provides the same sequential guarantees as Kotlin's Mutex.
 *
 * @example
 * ```typescript
 * const scheduler = new Scheduler();
 * scheduler.setComputeBudget(Budget.adaptive(BudgetProfile.BALANCED));
 *
 * const result = await scheduler.scheduleTask(
 *   TaskPriority.HIGH,
 *   WorkloadId.TEXT,
 *   () => edgeVeda.generate(prompt)
 * );
 * ```
 */

import {
  type EdgeVedaBudget,
  type MeasuredBaseline,
  type BudgetViolation,
  BudgetProfile,
  BudgetConstraint,
  WorkloadPriority,
  WorkloadId,
  Budget,
} from './Budget';
import { LatencyTracker } from './LatencyTracker';
import { BatteryDrainTracker } from './BatteryDrainTracker';
import { ThermalMonitor } from './ThermalMonitor';
import { ResourceMonitor } from './ResourceMonitor';

// =============================================================================
// Supporting Types
// =============================================================================

/** Task priority levels. */
export enum TaskPriority {
  LOW = 0,
  NORMAL = 1,
  HIGH = 2,
}

/** Task execution status. */
export enum TaskStatus {
  QUEUED = 'queued',
  RUNNING = 'running',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
  FAILED = 'failed',
}

/** Handle for a scheduled task. */
export interface TaskHandle {
  /** Unique task identifier. */
  id: string;
  /** Task priority. */
  priority: TaskPriority;
  /** Workload type. */
  workload: WorkloadId;
  /** Current status. */
  status: TaskStatus;
}

/** Queue status snapshot. */
export interface QueueStatus {
  /** Number of tasks currently queued. */
  queuedTasks: number;
  /** Number of tasks currently running. */
  runningTasks: number;
  /** Total number of completed tasks. */
  completedTasks: number;
  /** Number of high-priority tasks in queue. */
  highPriorityCount: number;
  /** Number of normal-priority tasks in queue. */
  normalPriorityCount: number;
  /** Number of low-priority tasks in queue. */
  lowPriorityCount: number;
}

// =============================================================================
// Internal PriorityQueue
// =============================================================================

interface QueuedItem {
  id: string;
  priority: TaskPriority;
}

class PriorityQueue {
  private items: QueuedItem[] = [];

  get count(): number {
    return this.items.length;
  }

  get isEmpty(): boolean {
    return this.items.length === 0;
  }

  enqueue(id: string, priority: TaskPriority): void {
    this.items.push({ id, priority });
    // Sort descending by priority value (HIGH=2 first)
    this.items.sort((a, b) => b.priority - a.priority);
  }

  dequeue(): string | undefined {
    if (this.items.length === 0) return undefined;
    return this.items.shift()!.id;
  }

  removeTask(id: string): void {
    this.items = this.items.filter((item) => item.id !== id);
  }

  countByPriority(priority: TaskPriority): number {
    return this.items.filter((item) => item.priority === priority).length;
  }
}

// =============================================================================
// UUID Helper
// =============================================================================

let _nextId = 0;

function generateTaskId(): string {
  _nextId += 1;
  const timestamp = Date.now().toString(36);
  const random = Math.random().toString(36).substring(2, 8);
  return `task_${timestamp}_${random}_${_nextId}`;
}

// =============================================================================
// Scheduler
// =============================================================================

/**
 * Priority-based task scheduler with budget enforcement.
 *
 * Manages concurrent inference workloads, enforces EdgeVedaBudget constraints,
 * and emits BudgetViolation events when limits are exceeded.
 *
 * Single-threaded JS means no mutex is needed — the API matches Swift/Kotlin
 * for cross-platform consistency.
 */
export class Scheduler {
  private readonly latencyTracker: LatencyTracker;
  private readonly batteryTracker: BatteryDrainTracker;
  private readonly thermalMonitor: ThermalMonitor;
  private readonly resourceMonitor: ResourceMonitor;

  private readonly taskQueue = new PriorityQueue();
  private budget: EdgeVedaBudget | undefined;
  private readonly workloadRegistry = new Map<WorkloadId, WorkloadPriority>();

  private _measuredBaseline: MeasuredBaseline | undefined;
  private warmUpComplete = false;
  private readonly warmUpThreshold = 20;

  private readonly violationListeners = new Map<
    string,
    (violation: BudgetViolation) => void
  >();
  private nextListenerId = 0;

  constructor(
    latencyTracker?: LatencyTracker,
    batteryTracker?: BatteryDrainTracker,
    thermalMonitor?: ThermalMonitor,
    resourceMonitor?: ResourceMonitor
  ) {
    this.latencyTracker = latencyTracker ?? new LatencyTracker();
    this.batteryTracker = batteryTracker ?? new BatteryDrainTracker();
    this.thermalMonitor = thermalMonitor ?? new ThermalMonitor();
    this.resourceMonitor = resourceMonitor ?? new ResourceMonitor();
  }

  // ---------------------------------------------------------------------------
  // Budget Management
  // ---------------------------------------------------------------------------

  /**
   * Set the compute budget for task execution.
   *
   * If the budget uses an adaptive profile, it will be resolved after warm-up
   * (20+ task samples). Until then, no budget enforcement occurs.
   */
  setComputeBudget(budget: EdgeVedaBudget): void {
    this.budget = budget;

    if (budget.adaptiveProfile !== undefined) {
      console.log(
        `[EdgeVeda.Scheduler] Adaptive budget set: ${budget.adaptiveProfile}. Warming up...`
      );
    }
  }

  /** Get the current compute budget. */
  getComputeBudget(): EdgeVedaBudget | undefined {
    return this.budget;
  }

  /**
   * Get the measured baseline after warm-up completes.
   *
   * Returns undefined if warm-up hasn't completed yet.
   */
  getMeasuredBaseline(): MeasuredBaseline | undefined {
    return this._measuredBaseline;
  }

  // ---------------------------------------------------------------------------
  // Task Scheduling
  // ---------------------------------------------------------------------------

  /**
   * Schedule a task with the specified priority.
   *
   * Tasks are queued and executed in priority order. High-priority tasks
   * run before normal and low-priority tasks. Budget constraints are checked
   * before each execution; violations are emitted but do not block execution.
   *
   * @param priority Task priority (HIGH, NORMAL, or LOW)
   * @param workload Workload type (TEXT or VISION)
   * @param task Async function to execute
   * @returns Result of the task execution
   */
  async scheduleTask<T>(
    priority: TaskPriority,
    workload: WorkloadId,
    task: () => Promise<T> | T
  ): Promise<T> {
    const taskId = generateTaskId();

    console.log(
      `[EdgeVeda.Scheduler] Task scheduled: ${taskId} priority=${TaskPriority[priority]}`
    );

    const startTime = Date.now();

    // Check budget before execution
    this.checkBudgetConstraints();

    // Execute task
    try {
      const result = await task();

      // Record latency
      const durationMs = Date.now() - startTime;
      this.latencyTracker.record(durationMs);

      // Update warm-up status
      if (!this.warmUpComplete && this.latencyTracker.sampleCount >= this.warmUpThreshold) {
        this.completeWarmUp();
      }

      console.log(
        `[EdgeVeda.Scheduler] Task completed: ${taskId} duration=${durationMs}ms`
      );
      return result;
    } catch (error) {
      console.error(
        `[EdgeVeda.Scheduler] Task failed: ${taskId} error=${error}`
      );
      throw error;
    }
  }

  /**
   * Cancel a scheduled task by ID.
   *
   * Only queued tasks can be cancelled. Running tasks cannot be cancelled.
   */
  cancelTask(taskId: string): void {
    this.taskQueue.removeTask(taskId);
  }

  /** Get current queue status. */
  getQueueStatus(): QueueStatus {
    return {
      queuedTasks: this.taskQueue.count,
      runningTasks: 0, // Simplified: single-threaded JS
      completedTasks: this.latencyTracker.sampleCount,
      highPriorityCount: this.taskQueue.countByPriority(TaskPriority.HIGH),
      normalPriorityCount: this.taskQueue.countByPriority(TaskPriority.NORMAL),
      lowPriorityCount: this.taskQueue.countByPriority(TaskPriority.LOW),
    };
  }

  // ---------------------------------------------------------------------------
  // Workload Management
  // ---------------------------------------------------------------------------

  /** Register a workload with its priority for degradation policy. */
  registerWorkload(workload: WorkloadId, priority: WorkloadPriority): void {
    this.workloadRegistry.set(workload, priority);
  }

  // ---------------------------------------------------------------------------
  // Violation Callbacks
  // ---------------------------------------------------------------------------

  /**
   * Register a callback for budget violation events.
   *
   * @param callback Called when a budget constraint is violated.
   * @returns Listener ID to use for removal via removeViolationListener().
   */
  onBudgetViolation(callback: (violation: BudgetViolation) => void): string {
    const id = `violation_${this.nextListenerId++}`;
    this.violationListeners.set(id, callback);
    return id;
  }

  /** Remove a budget violation listener. */
  removeViolationListener(id: string): void {
    this.violationListeners.delete(id);
  }

  // ---------------------------------------------------------------------------
  // Accessors for sub-components (useful for external telemetry / testing)
  // ---------------------------------------------------------------------------

  /** Get the internal LatencyTracker instance. */
  getLatencyTracker(): LatencyTracker {
    return this.latencyTracker;
  }

  /** Get the internal BatteryDrainTracker instance. */
  getBatteryTracker(): BatteryDrainTracker {
    return this.batteryTracker;
  }

  /** Get the internal ThermalMonitor instance. */
  getThermalMonitor(): ThermalMonitor {
    return this.thermalMonitor;
  }

  /** Get the internal ResourceMonitor instance. */
  getResourceMonitor(): ResourceMonitor {
    return this.resourceMonitor;
  }

  // ---------------------------------------------------------------------------
  // Private - Warm-up
  // ---------------------------------------------------------------------------

  private completeWarmUp(): void {
    if (this.budget === undefined) return;
    const profile = this.budget.adaptiveProfile;
    if (profile === undefined) return;

    const baseline: MeasuredBaseline = {
      measuredP95Ms: this.latencyTracker.p95,
      measuredDrainPerTenMin: this.batteryTracker.currentDrainRate,
      currentThermalState: this.thermalMonitor.currentLevel,
      currentRssMb: this.resourceMonitor.currentRssMb,
      sampleCount: this.latencyTracker.sampleCount,
      measuredAt: new Date(),
    };

    const resolvedBudget = Budget.resolve(profile, baseline);

    this._measuredBaseline = baseline;
    this.budget = resolvedBudget;
    this.warmUpComplete = true;

    console.log(
      `[EdgeVeda.Scheduler] Warm-up complete: ${Budget.baselineToString(baseline)}`
    );
    console.log(
      `[EdgeVeda.Scheduler] Resolved budget: ${Budget.toString(resolvedBudget)}`
    );
  }

  // ---------------------------------------------------------------------------
  // Private - Budget Enforcement
  // ---------------------------------------------------------------------------

  private checkBudgetConstraints(): void {
    const currentBudget = this.budget;
    if (currentBudget === undefined) return;
    if (!this.warmUpComplete) return;

    // Check p95 latency
    if (currentBudget.p95LatencyMs !== undefined) {
      const currentP95 = this.latencyTracker.p95;
      if (currentP95 > currentBudget.p95LatencyMs) {
        this.handleViolation(
          BudgetConstraint.P95_LATENCY,
          currentP95,
          currentBudget.p95LatencyMs
        );
      }
    }

    // Check battery drain
    if (currentBudget.batteryDrainPerTenMinutes !== undefined) {
      const currentDrain = this.batteryTracker.currentDrainRate;
      if (currentDrain !== undefined && currentDrain > currentBudget.batteryDrainPerTenMinutes) {
        this.handleViolation(
          BudgetConstraint.BATTERY_DRAIN,
          currentDrain,
          currentBudget.batteryDrainPerTenMinutes
        );
      }
    }

    // Check thermal level
    if (currentBudget.maxThermalLevel !== undefined) {
      const currentThermal = this.thermalMonitor.currentLevel;
      if (currentThermal > currentBudget.maxThermalLevel) {
        this.handleViolation(
          BudgetConstraint.THERMAL_LEVEL,
          currentThermal,
          currentBudget.maxThermalLevel
        );
      }
    }

    // Check memory ceiling (observe-only)
    if (currentBudget.memoryCeilingMb !== undefined) {
      const currentMemory = this.resourceMonitor.currentRssMb;
      if (currentMemory > currentBudget.memoryCeilingMb) {
        this.handleViolation(
          BudgetConstraint.MEMORY_CEILING,
          currentMemory,
          currentBudget.memoryCeilingMb
        );
      }
    }
  }

  private handleViolation(
    constraint: BudgetConstraint,
    currentValue: number,
    budgetValue: number
  ): void {
    const mitigation = this.attemptMitigation(constraint);

    const violation: BudgetViolation = {
      constraint,
      currentValue,
      budgetValue,
      mitigation,
      timestamp: new Date(),
      mitigated: false,
      observeOnly: constraint === BudgetConstraint.MEMORY_CEILING,
    };

    this.emitViolation(violation);
  }

  private attemptMitigation(constraint: BudgetConstraint): string {
    switch (constraint) {
      case BudgetConstraint.P95_LATENCY:
        return 'Reduce inference frequency';
      case BudgetConstraint.BATTERY_DRAIN:
        return 'Lower model quality';
      case BudgetConstraint.THERMAL_LEVEL:
        return 'Pause high-priority workloads';
      case BudgetConstraint.MEMORY_CEILING:
        return 'Observe only - cannot reduce model memory';
    }
  }

  private emitViolation(violation: BudgetViolation): void {
    console.warn(
      `[EdgeVeda.Scheduler] ⚠️ Budget Violation: ${Budget.violationToString(violation)}`
    );

    for (const listener of this.violationListeners.values()) {
      try {
        listener(violation);
      } catch (error) {
        console.warn(
          `[EdgeVeda.Scheduler] Error in violation listener: ${error}`
        );
      }
    }
  }
}

File: kotlin/src/main/kotlin/com/edgeveda/sdk/Scheduler.kt
package com.edgeveda.sdk

import android.util.Log
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.Date
import java.util.UUID
import kotlin.coroutines.cancellation.CancellationException

/**
 * Priority-based task scheduler with budget enforcement.
 *
 * The Scheduler manages concurrent inference workloads, enforces [EdgeVedaBudget]
 * constraints, and emits [BudgetViolation] events when limits are exceeded.
 *
 * Thread-safe via [Mutex] (Kotlin equivalent of Swift actor isolation).
 *
 * Example:
 * ```kotlin
 * val scheduler = Scheduler()
 * scheduler.setComputeBudget(EdgeVedaBudget.adaptive(BudgetProfile.BALANCED))
 *
 * val result = scheduler.scheduleTask(
 *     priority = TaskPriority.HIGH,
 *     workload = WorkloadId.TEXT
 * ) {
 *     edgeVeda.generate(prompt)
 * }
 * ```
 */
class Scheduler(
    private val latencyTracker: LatencyTracker = LatencyTracker(),
    private val batteryTracker: BatteryDrainTracker = BatteryDrainTracker(),
    private val thermalMonitor: ThermalMonitor = ThermalMonitor(),
    private val resourceMonitor: ResourceMonitor = ResourceMonitor()
) {
    private val mutex = Mutex()
    private val taskQueue = PriorityQueue()
    private var budget: EdgeVedaBudget? = null
    private val workloadRegistry = mutableMapOf<WorkloadId, WorkloadPriority>()

    private var measuredBaseline: MeasuredBaseline? = null
    private var warmUpComplete = false
    private val warmUpThreshold = 20 // samples needed

    private val violationListeners = mutableMapOf<String, (BudgetViolation) -> Unit>()

    companion object {
        private const val TAG = "EdgeVeda.Scheduler"
    }

    // -------------------------------------------------------------------
    // Budget Management
    // -------------------------------------------------------------------

    /**
     * Set the compute budget for task execution.
     *
     * If the budget uses an adaptive profile, it will be resolved after warm-up
     * (20+ task samples). Until then, no budget enforcement occurs.
     */
    suspend fun setComputeBudget(budget: EdgeVedaBudget) = mutex.withLock {
        this.budget = budget

        if (budget.adaptiveProfile != null) {
            Log.i(TAG, "Adaptive budget set: ${budget.adaptiveProfile.name}. Warming up...")
        }
    }

    /** Get the current compute budget. */
    suspend fun getComputeBudget(): EdgeVedaBudget? = mutex.withLock { budget }

    /**
     * Get the measured baseline after warm-up completes.
     *
     * Returns null if warm-up hasn't completed yet.
     */
    suspend fun getMeasuredBaseline(): MeasuredBaseline? = mutex.withLock { measuredBaseline }

    // -------------------------------------------------------------------
    // Task Scheduling
    // -------------------------------------------------------------------

    /**
     * Schedule a task with the specified priority.
     *
     * Tasks are queued and executed in priority order. High-priority tasks
     * run before normal and low-priority tasks. Budget constraints are checked
     * before each execution; violations are emitted but do not block execution.
     *
     * @param priority Task priority (HIGH, NORMAL, or LOW)
     * @param workload Workload type (TEXT or VISION)
     * @param task Suspend function to execute
     * @return Result of the task execution
     * @throws Exception Any error from the task execution
     */
    suspend fun <T> scheduleTask(
        priority: TaskPriority,
        workload: WorkloadId,
        task: suspend () -> T
    ): T {
        val taskId = UUID.randomUUID().toString()
        val taskHandle = TaskHandle(
            id = taskId,
            priority = priority,
            workload = workload,
            status = TaskStatus.QUEUED
        )

        Log.d(TAG, "Task scheduled: $taskId priority=${priority.name}")

        val startTime = System.currentTimeMillis()

        // Check budget before execution
        checkBudgetConstraints()

        // Execute task
        try {
            val result = task()

            // Record latency
            val durationMs = (System.currentTimeMillis() - startTime).toDouble()
            latencyTracker.record(durationMs)

            // Update warm-up status
            val currentSampleCount = latencyTracker.sampleCount()
            if (!warmUpComplete && currentSampleCount >= warmUpThreshold) {
                completeWarmUp()
            }

            Log.d(TAG, "Task completed: $taskId duration=${durationMs.toLong()}ms")
            return result
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "Task failed: $taskId error=${e.message}")
            throw e
        }
    }

    /**
     * Cancel a scheduled task by ID.
     *
     * Only queued tasks can be cancelled. Running tasks cannot be cancelled.
     */
    suspend fun cancelTask(taskId: String) = mutex.withLock {
        taskQueue.removeTask(taskId)
    }

    /** Get current queue status. */
    suspend fun getQueueStatus(): QueueStatus = mutex.withLock {
        QueueStatus(
            queuedTasks = taskQueue.count,
            runningTasks = 0, // Simplified: coroutine-based execution
            completedTasks = latencyTracker.sampleCount(),
            highPriorityCount = taskQueue.countByPriority(TaskPriority.HIGH),
            normalPriorityCount = taskQueue.countByPriority(TaskPriority.NORMAL),
            lowPriorityCount = taskQueue.countByPriority(TaskPriority.LOW)
        )
    }

    // -------------------------------------------------------------------
    // Workload Management
    // -------------------------------------------------------------------

    /** Register a workload with its priority for degradation policy. */
    suspend fun registerWorkload(workload: WorkloadId, priority: WorkloadPriority) = mutex.withLock {
        workloadRegistry[workload] = priority
    }

    // -------------------------------------------------------------------
    // Violation Callbacks
    // -------------------------------------------------------------------

    /**
     * Register a callback for budget violation events.
     *
     * @param callback Called when a budget constraint is violated
     * @return Listener ID to use for removal via [removeViolationListener]
     */
    suspend fun onBudgetViolation(callback: (BudgetViolation) -> Unit): String = mutex.withLock {
        val id = UUID.randomUUID().toString()
        violationListeners[id] = callback
        id
    }

    /** Remove a budget violation listener. */
    suspend fun removeViolationListener(id: String) = mutex.withLock {
        violationListeners.remove(id)
    }

    // -------------------------------------------------------------------
    // Private - Warm-up
    // -------------------------------------------------------------------

    private suspend fun completeWarmUp() {
        val currentBudget = mutex.withLock { budget } ?: return
        val profile = currentBudget.adaptiveProfile ?: return

        val baseline = MeasuredBaseline(
            measuredP95Ms = latencyTracker.p95(),
            measuredDrainPerTenMin = batteryTracker.currentDrainRate(),
            currentThermalState = thermalMonitor.currentLevel(),
            currentRssMb = resourceMonitor.currentRssMb(),
            sampleCount = latencyTracker.sampleCount(),
            measuredAt = Date()
        )

        val resolvedBudget = EdgeVedaBudget.resolve(profile, baseline)

        mutex.withLock {
            this.measuredBaseline = baseline
            this.budget = resolvedBudget
            this.warmUpComplete = true
        }

        Log.i(TAG, "Warm-up complete: $baseline")
        Log.i(TAG, "Resolved budget: $resolvedBudget")
    }

    // -------------------------------------------------------------------
    // Private - Budget Enforcement
    // -------------------------------------------------------------------

    private suspend fun checkBudgetConstraints() {
        val currentBudget: EdgeVedaBudget
        val isWarm: Boolean
        mutex.withLock {
            currentBudget = budget ?: return
            isWarm = warmUpComplete
        }
        if (!isWarm) return

        // Check p95 latency
        currentBudget.p95LatencyMs?.let { maxP95 ->
            val currentP95 = latencyTracker.p95()
            if (currentP95 > maxP95.toDouble()) {
                handleViolation(
                    constraint = BudgetConstraint.P95_LATENCY,
                    currentValue = currentP95,
                    budgetValue = maxP95.toDouble()
                )
            }
        }

        // Check battery drain
        currentBudget.batteryDrainPerTenMinutes?.let { maxDrain ->
            batteryTracker.currentDrainRate()?.let { currentDrain ->
                if (currentDrain > maxDrain) {
                    handleViolation(
                        constraint = BudgetConstraint.BATTERY_DRAIN,
                        currentValue = currentDrain,
                        budgetValue = maxDrain
                    )
                }
            }
        }

        // Check thermal level
        currentBudget.maxThermalLevel?.let { maxThermal ->
            val currentThermal = thermalMonitor.currentLevel()
            if (currentThermal > maxThermal) {
                handleViolation(
                    constraint = BudgetConstraint.THERMAL_LEVEL,
                    currentValue = currentThermal.toDouble(),
                    budgetValue = maxThermal.toDouble()
                )
            }
        }

        // Check memory ceiling (observe-only)
        currentBudget.memoryCeilingMb?.let { maxMemory ->
            val currentMemory = resourceMonitor.currentRssMb()
            if (currentMemory > maxMemory.toDouble()) {
                handleViolation(
                    constraint = BudgetConstraint.MEMORY_CEILING,
                    currentValue = currentMemory,
                    budgetValue = maxMemory.toDouble()
                )
            }
        }
    }

    private suspend fun handleViolation(
        constraint: BudgetConstraint,
        currentValue: Double,
        budgetValue: Double
    ) {
        val mitigation = attemptMitigation(constraint)

        val violation = BudgetViolation(
            constraint = constraint,
            currentValue = currentValue,
            budgetValue = budgetValue,
            mitigation = mitigation,
            timestamp = Date(),
            mitigated = false,
            observeOnly = constraint == BudgetConstraint.MEMORY_CEILING
        )

        emitViolation(violation)
    }

    private fun attemptMitigation(constraint: BudgetConstraint): String {
        return when (constraint) {
            BudgetConstraint.P95_LATENCY -> "Reduce inference frequency"
            BudgetConstraint.BATTERY_DRAIN -> "Lower model quality"
            BudgetConstraint.THERMAL_LEVEL -> "Pause high-priority workloads"
            BudgetConstraint.MEMORY_CEILING -> "Observe only - cannot reduce model memory"
        }
    }

    private suspend fun emitViolation(violation: BudgetViolation) {
        Log.w(TAG, "⚠️ Budget Violation: $violation")

        val listeners = mutex.withLock { violationListeners.values.toList() }
        for (listener in listeners) {
            try {
                listener(violation)
            } catch (e: Exception) {
                Log.w(TAG, "Error in violation listener: ${e.message}")
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Supporting Types
// ---------------------------------------------------------------------------

/** Handle for a scheduled task. */
data class TaskHandle(
    val id: String,
    val priority: TaskPriority,
    val workload: WorkloadId,
    val status: TaskStatus
)

/** Task priority levels. */
enum class TaskPriority(val value: Int) {
    LOW(0),
    NORMAL(1),
    HIGH(2);

    companion object {
        fun fromValue(value: Int): TaskPriority = entries.first { it.value == value }
    }
}

/** Task execution status. */
enum class TaskStatus {
    QUEUED,
    RUNNING,
    COMPLETED,
    CANCELLED,
    FAILED
}

/** Queue status snapshot. */
data class QueueStatus(
    val queuedTasks: Int,
    val runningTasks: Int,
    val completedTasks: Int,
    val highPriorityCount: Int,
    val normalPriorityCount: Int,
    val lowPriorityCount: Int
) {
    override fun toString(): String =
        "QueueStatus(queued=$queuedTasks, running=$runningTasks, " +
            "completed=$completedTasks, high=$highPriorityCount, " +
            "normal=$normalPriorityCount, low=$lowPriorityCount)"
}

// ---------------------------------------------------------------------------
// PriorityQueue (internal)
// ---------------------------------------------------------------------------

/** Internal priority queue for scheduled tasks. */
internal class PriorityQueue {
    private data class QueuedItem(
        val id: String,
        val priority: TaskPriority
    )

    private val items = mutableListOf<QueuedItem>()

    val count: Int get() = items.size
    val isEmpty: Boolean get() = items.isEmpty()

    fun enqueue(id: String, priority: TaskPriority) {
        items.add(QueuedItem(id, priority))
        items.sortByDescending { it.priority.value }
    }

    fun dequeue(): String? {
        if (items.isEmpty()) return null
        return items.removeFirst().id
    }

    fun removeTask(id: String) {
        items.removeAll { it.id == id }
    }

    fun countByPriority(priority: TaskPriority): Int =
        items.count { it.priority == priority }
}